### PR TITLE
Ananth ch16 KL divergence cross-entropy cost section added

### PR DESCRIPTION
**Pull request by ananthagarwal**
First draft for a section on KL divergence and how it relates to cross-entropy cost.

### CODE REVIEW

**Comment by ananthagarwal** [comment]
Hey Sam, I added changes based on our conversation from Tuesday. Let me know what you think. There are two things I want to mention:
1. I'm not sure whether the bottom of the section is the best place to add my content. For now I have a sentence above the "Gradient of the Cross Entropy Cost" section that refers readers to it.
2. In the cross-entropy equations in the first half of the section, X_i is used instead of x_i. I believe it should be x_i to refer to specific rows within X. Let me know if this is the case and I can make the fixes.

Thank you for your time and have a great weekend!

**Comment by junseo-park** [comment]
Fixed a merge conflict with master from a wording change I made!

File: notebooks/ch16/classification_cost.md [hunk 1]
```diff
@@ -195,4 +205,55 @@ The surprisingly simple gradient expression allows us to fit a logistic model to
...
+
+The KL divergence between two distributions $P$ and $\hat{P_\theta}$ for a single data point $(x, y)$ is given by:
+
+$$D(P || \hat{P_\theta}) = \sum_{k=0}^{K-1} P(y = k | x) \ln \left(\frac{P(y = k | x)}{\hat{P_\theta}(y = k | x)}\right)$$
+
+$K$ represents the number of possible outcomes, so in the case of binary classification $K = 2$. 
+
+The best $\theta$ minimizes the average KL divergence of the entire dataset of $n$ points:
+
+$$\displaystyle\arg \min_{\substack{\theta}} \frac{1}{n} \sum_{i=1}^{n} \sum_{k=0}^{K-1} P(y_i = k | x_i) \ln \left(\frac{P(y_i = k | x_i)}{\hat{P_\theta}(y_i = k | x_i)}\right)$$
```
Comment (SamLau95) [inline notebooks/ch16/classification_cost.md]: Since we aren't really using this equation right now, why not move it below to the derivation part of the section?

File: notebooks/ch16/classification_cost.md [hunk 2]
```diff
@@ -195,4 +205,55 @@ The surprisingly simple gradient expression allows us to fit a logistic model to
...
+
+Average KL divergence can be interpreted as the average log difference between the two distributions $P$ and $\hat{P_\theta}$ weighted by $P$. The structure of the equation has two important implications:
```
Comment (SamLau95) [inline notebooks/ch16/classification_cost.md]: The first sentence here won't have much meaning or usefulness for the first-time reader. Can we either explain it more or omit it?

File: notebooks/ch16/classification_cost.md [hunk 3]
```diff
@@ -195,4 +205,55 @@ The surprisingly simple gradient expression allows us to fit a logistic model to
...
+
+* KL divergence does not penalize mismatch for rare events with respect to $P$. If the model predicts a high probability for an event that is actually rare, then both $P(k)$ and $\ln \left(\frac{P(k)}{\hat{P}(k)}\right)$ are low so the divergence is also low. However, if the model predicts a low probability for an event that is actually common, then the divergence is high. 
+* KL divergence is not symmetric: $$D(P || \hat{P_\theta}) \neq D(\hat{P_\theta} || P)$$ In other words, the divergence of $\hat{P_\theta}$ from $P$ is not the same as the divergence of $P$ from $\hat{P_\theta}$.
```
Comment (SamLau95) [inline notebooks/ch16/classification_cost.md]: Can we end this with a sentence saying something like: "Because we use $\hat{P_\theta}$ to approximate $P$, we use $ D(P || \hat{P_\theta}) $."

File: notebooks/ch16/classification_cost.md [hunk 4]
```diff
@@ -195,4 +205,55 @@ The surprisingly simple gradient expression allows us to fit a logistic model to
...
+
+We can deduce that a logistic regression model that accurately predicts common events has a lower divergence from the real distribution than does a model that accurately predicts rare events but varies widely on common events. 
+
+### Deriving Cross-Entropy Cost from KL Divergence
+
+The logarithm of the ratio of probabilities in the average KL divergence equation hints back to our study of the cross-entropy cost equation. In fact, minimizing average KL divergence is equivalent to minimizing average cross-entropy cost. The rest of this section walks through an algebraic derivation of cross-entropy cost from average KL divergence.
```
Comment (SamLau95) [inline notebooks/ch16/classification_cost.md]: Since I think we should mention that minimizing KL divergence is equivalent to minimizing cross-entropy earlier in the section, the second sentence should be reworded or omitted.

nit: Let's reword the last sentence to "We will demonstrate this equivalence by deriving the cross-entropy cost from the KL divergence mathematically."

File: notebooks/ch16/classification_cost.md [hunk 5]
```diff
@@ -195,4 +205,55 @@ The surprisingly simple gradient expression allows us to fit a logistic model to
...
+
+Using properties of logarithms, we can rewrite the weighted log ratio:
+$$P(y_i = k | x_i) \ln \left(\frac{P(y_i = k | x_i)}{\hat{P_\theta}(y_i = k | x_i)}\right) = P(y_i = k | x_i) \ln P(y_i = k | x_i) - P(y_i = k | x_i) \ln \hat{P_\theta}(y_i = k | x_i)$$
+
+Note that since the first term produced doesn't depend on $\theta$, it doesn't affect $\displaystyle\arg \min_{\substack{\theta}}$ and can be removed from the equation. The resulting equation is the cross-entropy cost of the model $\hat{P_\theta}$:
```
Comment (SamLau95) [inline notebooks/ch16/classification_cost.md]: nit: Remove "produced"

File: notebooks/ch16/classification_cost.md [hunk 6]
```diff
@@ -195,4 +205,55 @@ The surprisingly simple gradient expression allows us to fit a logistic model to
...
+
+$$\displaystyle\arg \min_{\substack{\theta}} \frac{1}{n} \sum_{i=1}^{n} \sum_{k=0}^{K-1} -P(y_i = k | x_i) \ln \hat{P_\theta}(y_i = k | x_i)$$
+
+
+We can now substitute equivalent terms to further simplify the equation to its more familiar form. Expanding the innermost summation over $K$ results in the following:
+
+$$\displaystyle\arg \min_{\substack{\theta}} \frac{1}{n} \sum_{i=1}^{n} - P(y_i = 0 | x_i) \ln \hat{P_\theta}(y_i = 0 | x_i) - P(y_i = 1 | x_i) \ln \hat{P_\theta}(y_i = 1 | x_i)$$
+
+The label $y_i$ is either 0 or 1. Since it is a known value, the probability that $y_i = 1$, $P(y_i = 1 | x_i)$, is equal to $y_i$ and $P(y_i = 0 | x_i)$ is equal to $1 - y_i$. The model's probability distribution $\hat{P_\theta}$ is given by the output of the sigmoid function from the beginning of the section. Putting it all together, we arrive at the cross-entropy cost equation:
```
Comment (SamLau95) [inline notebooks/ch16/classification_cost.md]: nit: Replace "Putting it all together" with "After making these substitutions". The phrase "putting it all together" is a bit unclear since we don't know what "it" refers to.

File: notebooks/ch16/classification_cost.md [hunk 7]
```diff
@@ -195,4 +205,55 @@ The surprisingly simple gradient expression allows us to fit a logistic model to
...
+
+$$ \displaystyle\arg \min_{\substack{\theta}} \frac{1}{n} \sum_i \left(- y_i \ln (f_\hat{\theta}(x_i)) - (1 - y_i) \ln (1 - f_\hat{\theta}(x_i) \right) $$
+
+### Summary
+
+Minimizing average KL divergence also minimizes average cross-entropy cost. We can reduce the divergence of logistic regression models by selecting features that accurately classify common data. For a more in-depth study of cross-entropy and KL divergence, interested readers are encouraged to consult resources on information theory.
```
Comment (SamLau95) [inline notebooks/ch16/classification_cost.md]: I don't think this last sentence is that helpful since there are clearly many resources on information theory. I propose we either omit this last sentence or point them to specific resources that we think are particularly helpful.

File: notebooks/ch16/classification_cost.md [hunk 8]
```diff
@@ -195,4 +205,55 @@ The surprisingly simple gradient expression allows us to fit a logistic model to
 
 ## Summary
 
-The cross entropy cost that we use to fit the logistic model. Since it is a convex function, we use gradient descent to fit the model to the cost. We now have the necessary components of logistic regression: the model, cost function, and minimization procedure.
+The cost function for logistic models is based on cross-entropy loss. Since it is a convex function, we use gradient descent to fit the model to the cost. We now have the necessary components of logistic regression: the model, cost function, and minimization procedure.
+
+## KL Divergence and Cross-Entropy Cost
```
[the remark this answers is not in the document]
Comment (ananthagarwal) [reply notebooks/ch16/classification_cost.md]: Yeah I think it sounds like a good idea! It helps to divide up topics better and be less intimidating for first-time readers.

File: notebooks/ch16/classification_cost.md [hunk 9]
```diff
@@ -195,4 +205,55 @@ The surprisingly simple gradient expression allows us to fit a logistic model to
...
+
+$$ \hat{P_\theta}(y = 1 | x) \approx P(y = 1 | x) $$
+
+**Kullback–Leibler (KL) divergence** provides a measure of the difference between two probability distributions. It reveals how imprecise the probability distribution output by a logistic regression model is relative to the data. KL-divergence, also called relative entropy, has its roots in information theory (beyond the scope of DS 100).
+
+The KL divergence between two distributions $P$ and $\hat{P_\theta}$ for a single data point $(x, y)$ is given by:
+
+$$D(P || \hat{P_\theta}) = \sum_{k=0}^{K-1} P(y = k | x) \ln \left(\frac{P(y = k | x)}{\hat{P_\theta}(y = k | x)}\right)$$
+
+$K$ represents the number of possible outcomes, so in the case of binary classification $K = 2$.
```
[the remark this answers is not in the document]
Comment (ananthagarwal) [reply notebooks/ch16/classification_cost.md]: Sounds good

File: notebooks/ch16/classification_cost.md [hunk 10]
```diff
@@ -195,4 +205,55 @@ The surprisingly simple gradient expression allows us to fit a logistic model to
...
 ## Summary
 
-The cross entropy cost that we use to fit the logistic model. Since it is a convex function, we use gradient descent to fit the model to the cost. We now have the necessary components of logistic regression: the model, cost function, and minimization procedure.
+The cost function for logistic models is based on cross-entropy loss. Since it is a convex function, we use gradient descent to fit the model to the cost. We now have the necessary components of logistic regression: the model, cost function, and minimization procedure.
+
+## KL Divergence and Cross-Entropy Cost
+
+### Defining KL Divergence
+
+As noted at the beginning of the section, the goal in our binary classification example is to closely approximate the real data distribution using a logistic model with parameters $\theta$:
```
[the remark this answers is not in the document]
Comment (ananthagarwal) [reply notebooks/ch16/classification_cost.md]: I agree, especially since we are making this a new section it will be helpful to provide more context. Originally I wasn't sure how much detail to add in order to prevent it from becoming too long, but this definitely works for the new section.

**Comment by ananthagarwal** [comment]
I created the new section for KL divergence and incorporated your comments. Let me know what you think!

**Comment by SamLau95** [comment]
Thanks for moving your work into a new section! Can you name the notebook `classification_cost_justification.ipynb` so that the notebook has the same prefix as the other notebooks in ch16? Also, I don't see a markdown file for your new notebook. Can you try regenerating the markdown with your rename?

File: notebooks/ch16/classification_cost.md [hunk 11]
```diff
@@ -195,4 +270,4 @@ The surprisingly simple gradient expression allows us to fit a logistic model to
 
 ## Summary
 
-The cross entropy cost that we use to fit the logistic model. Since it is a convex function, we use gradient descent to fit the model to the cost. We now have the necessary components of logistic regression: the model, cost function, and minimization procedure.
+The cost function for logistic models is based on cross-entropy loss. Since it is a convex function, we use gradient descent to fit the model to the cost. We now have the necessary components of logistic regression: the model, cost function, and minimization procedure.
```
Comment (SamLau95) [inline notebooks/ch16/classification_cost.md]: nit: The cost function isn't based on cross-entropy loss, it *is* cross-entropy loss. How about:

> Since the cross-entropy loss function is convex, we minimize this loss using gradient descent to fit logistic models to data.

File: notebooks/ch16/classification_cost_justification.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+$$ \hat{P_\theta}(y = 1 | x) \approx P(y = 1 | x) $$
+
+One commonly used metric to determine the "closeness" of two probability distributions is the Kullback–Leibler divergence, or KL divergence, which has its roots in information theory.
+
+## Defining Average KL Divergence
+
+KL Divergence reveals how imprecise the probability distribution output by a logistic model is relative to the data. For example, revisiting the basketball data, suppose we want to determine the probability that LeBron makes a shot 15 ft from the basket. KL divergence quantifies the difference between the probability output by our logistic model and the actual probability based on the dataset. The KL divergence for binary classification between two distributions $P$ and $\hat{P_\theta}$ for a single data point $(x, y)$ is given by:
```
Comment (SamLau95) [inline notebooks/ch16/classification_cost_justification.md]: The second sentence doesn't seem to serve much purpose in this paragraph since it doesn't mention probability distributions at all. Did you mean to explain it further?

Otherwise, we might want to reorder the sentences in this paragraph a bit:

> KL divergence quantifies the difference between the probability distribution computed by our logistic model and the actual distribution based on the dataset. Intuitively, it calculates how imprecisely the logistic model estimates the distribution of labels in data.
> 
> The KL divergence for binary classification between two distributions $P$ and $\hat{P_\theta}$ for a single data point $(x, y)$ is given by:

File: notebooks/ch16/classification_cost_justification.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+$$D(P || \hat{P_\theta}) = P(y = 0 | x) \ln \left(\frac{P(y = 0 | x)}{\hat{P_\theta}(y = 0 | x)}\right) + P(y = 1 | x) \ln \left(\frac{P(y = 1 | x)}{\hat{P_\theta}(y = 1 | x)}\right)$$
+
+KL divergence is not symmetric: $$D(P || \hat{P_\theta}) \neq D(\hat{P_\theta} || P)$$ In other words, the divergence of $\hat{P_\theta}$ from $P$ is not the same as the divergence of $P$ from $\hat{P_\theta}$. Since we use $\hat{P_\theta}$ to approximate $P$, we are concerned with $ D(P || \hat{P_\theta}) $.
```
Comment (SamLau95) [inline notebooks/ch16/classification_cost_justification.md]: nit: Missing period after first sentence.

File: notebooks/ch16/classification_cost_justification.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+KL divergence does not penalize mismatch for rare events with respect to $P$. If the model predicts a high probability for an event that is actually rare, then both $P(k)$ and $\ln \left(\frac{P(k)}{\hat{P}(k)}\right)$ are low so the divergence is also low. However, if the model predicts a low probability for an event that is actually common, then the divergence is high. We can deduce that a logistic model that accurately predicts common events has a lower divergence from $P$ than does a model that accurately predicts rare events but varies widely on common events. 
+
+## Deriving Cross-Entropy Cost from KL Divergence
+
+The structure of the average KL divergence equation for binary classification hints back to our study of the cross-entropy cost equation. The most important takeaway of this section is that minimizing average KL divergence is equivalent to minimizing average cross-entropy cost. We will demonstrate this equivalence by deriving the cross-entropy cost from the KL divergence mathematically.
```
Comment (SamLau95) [inline notebooks/ch16/classification_cost_justification.md]: We typically place important points at the beginning or end of paragraphs for emphasis. For example:

> The structure of the above average KL divergence equation contains some surface similarities with cross-entropy loss. We will now show with some algebraic manipulation that minimizing average KL divergence is in fact equivalent to minimizing average cross-entropy cost.

File: notebooks/ch16/classification_cost_justification.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+$$ \displaystyle\arg \min_{\substack{\theta}} \frac{1}{n} \sum_i \left(- y_i \ln (f_\hat{\theta}(x_i)) - (1 - y_i) \ln (1 - f_\hat{\theta}(x_i) \right) $$
+
+## Statistical justification for Cross-Entropy Cost
+
+The cross entropy cost also has fundamental underpinnings in statistics. Since the logistic regression model predicts probabilities, given a particular logistic model we can ask, "What is the probability that this model produced the set of observed outcomes $ y $?" We might naturally adjust the parameters of our model until the probability of drawing our dataset from the model is as high as possible. Although we will not prove it in this section, this procedure is equivalent to minimizing the cross entropy cost—this is the *maximum likelihood* statistical justification for the cross entropy cost.
+
+## Summary
+
+Average KL divergence can be interpreted as the average log difference between the two distributions $P$ and $\hat{P_\theta}$ weighted by $P$. Minimizing average KL divergence also minimizes average cross-entropy cost. We can reduce the divergence of logistic regression models by selecting features that accurately classify commonly occurring data.
```
Comment (SamLau95) [inline notebooks/ch16/classification_cost_justification.md]: nit: In the last sentence, we use KL divergence to select model parameters, not "features".

File: notebooks/ch16/classification_cost_justification.md (hunk 13 above)
Comment (ananthagarwal) [reply notebooks/ch16/classification_cost_justification.md]: I purposely omitted it since the Latex centers the equation.
